### PR TITLE
feat(backend): 单据批量确认 #5190

### PR DESCRIPTION
**Pull request by Maxwell-416258**

### CODE REVIEW

**Comment by tencentblueking-adm** [comment]
[CLA assistant check](https://bk-cla.bktencent.com:443/TencentBlueKing/blueking-dbm?pullRequest=5217) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://bk-cla.bktencent.com:443/TencentBlueKing/blueking-dbm?pullRequest=5217) before we can accept your contribution.<br/><hr/>**leifxu** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://bk-cla.bktencent.com:443/check/TencentBlueKing/blueking-dbm?pullRequest=5217) it.</sub>

File: dbm-ui/backend/ticket/serializers.py [hunk 1]
```diff
@@ -333,3 +333,22 @@ def validate(self, attrs):
...
+    todo_id = serializers.CharField(help_text=_("待办ID"))
+    params = serializers.JSONField(help_text=_("动作参数"))
+
+
+class BatchTodoOperateSerializer(serializers.Serializer):
+    """
+    批量待办处理
+    """
+
+    action = serializers.ChoiceField(choices=["APPROVE", "TERMINATE"], help_text=_("统一动作"))
```
Comment (iSecloud) [inline dbm-ui/backend/ticket/serializers.py]: 这里有个choice type： ActionType

File: dbm-ui/backend/ticket/views.py [hunk 2]
```diff
@@ -590,3 +591,48 @@ def fast_create_cloud_component(self, request, *args, **kwargs):
...
+    )
+    @action(methods=["POST"], detail=False, serializer_class=BatchTodoOperateSerializer)
+    def batch_process_todo(self, request, *args, **kwargs):
+        """
+        批量处理待办: 返回处理后的待办列表
+        """
+        # 使用 BatchTodoOperateSerializer 验证请求数据
+        serializer = self.get_serializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        validated_data = serializer.validated_data
```
Comment (iSecloud) [inline dbm-ui/backend/ticket/views.py]: 可以简写成一句：        validated_data = self.params_validate(self.get_serializer_class())

File: dbm-ui/backend/ticket/views.py [hunk 3]
```diff
@@ -590,3 +591,48 @@ def fast_create_cloud_component(self, request, *args, **kwargs):
...
+        # 批量处理待办操作
+        for operation in validated_data["operations"]:
+            ticket_id = operation["ticket_id"]
+            todo_id = operation["todo_id"]
+            params = operation["params"]
+
+            try:
+                ticket = Ticket.objects.get(id=ticket_id)
+                try:
+                    todo = ticket.todo_of_ticket.get(id=todo_id)
```
Comment (iSecloud) [inline dbm-ui/backend/ticket/views.py]: 这里看起来不需要ticket_id，也无需查询单据。让前端批量传todo id过来。通过Todo可以反查出ticket。

File: dbm-ui/backend/ticket/views.py [hunk 4]
```diff
@@ -590,3 +591,48 @@ def fast_create_cloud_component(self, request, *args, **kwargs):
...
+                    TodoActorFactory.actor(todo).process(request.user.username, action, params)
+                    results.append(todo)
+                except Todo.DoesNotExist:
+                    return Response(
```
Comment (iSecloud) [inline dbm-ui/backend/ticket/views.py]: 异常时不要返回response了，返回自定义异常。dbm和前端有统一捕获异常的机制

File: dbm-ui/backend/ticket/views.py [hunk 5]
```diff
@@ -590,3 +591,48 @@ def fast_create_cloud_component(self, request, *args, **kwargs):
...
+                        status=status.HTTP_404_NOT_FOUND,
+                    )
+            except Ticket.DoesNotExist:
+                return Response(
+                    {"detail": f"Ticket with id {ticket_id} does not exist."}, status=status.HTTP_404_NOT_FOUND
+                )
+
+        # 使用 TodoSerializer 序列化响应数据
+        response_serializer = TodoSerializer(results, many=True)
+        return Response(response_serializer.data, status=status.HTTP_200_OK)
```
Comment (iSecloud) [inline dbm-ui/backend/ticket/views.py]: 这里感觉直接：
```python
return Reponse(results)
```
就好了。
为什么再序列化一次，然后在得到序列化的data返回呢？

File: dbm-ui/backend/ticket/views.py [hunk 6]
```diff
@@ -590,3 +591,48 @@ def fast_create_cloud_component(self, request, *args, **kwargs):
...
+                try:
+                    todo = ticket.todo_of_ticket.get(id=todo_id)
+                    TodoActorFactory.actor(todo).process(request.user.username, action, params)
+                    results.append(todo)
+                except Todo.DoesNotExist:
+                    return Response(
+                        {"detail": f"Todo with id {todo_id} does not exist in ticket {ticket_id}."},
+                        status=status.HTTP_404_NOT_FOUND,
+                    )
+            except Ticket.DoesNotExist:
```
Comment (iSecloud) [inline dbm-ui/backend/ticket/views.py]: 还有这里提个建议，尽量不要在try except里面在嵌套try except。除非异常处理的逻辑不同。
比如对于此处，可以简写为：
```python
try:
      xxxx
except (Ticket.DoesNotExist, Todo.DoesNotExist):
      raise xxx  // 抛出自定义异常
```

还有对于这些模型存在类的前置检查，也可以放到serializer去做。
这里尽量只关注业务逻辑的异常情况，比如确认失败，比如todo后的run next flow失败等等。

File: dbm-ui/backend/ticket/serializers.py [hunk 7]
```diff
@@ -333,3 +334,30 @@ def validate(self, attrs):
...
+class BatchTodoOperation(serializers.Serializer):
+    """
+    批量待办操作
+    """
+
+    todo_id = serializers.CharField(help_text=_("待办ID"))
+    params = serializers.JSONField(help_text=_("动作参数"))
+
+    def validate(self, attrs):
+        try:
```
Comment (iSecloud) [inline dbm-ui/backend/ticket/serializers.py]: 这样是n+1查询，因为每个都会校验。

**Review by iSecloud — APPROVED** [review]

**Pull request closed without merging**